### ISSUE TITLE
Allow spacing, size, etc. to be configurable

### ISSUE THREAD

**Issue by jackhumbert** [jackhumbert/in_world_navigation#1]

**Comment by Thortok** [comment]
If/when you have some builds to test this, ping me and I'll give 'em a test for ya

**Comment by Thortok** [comment]
I know it's not actual settings yet, but are there any lines of code that I could edit some variables to play around with different settings?

**Comment by jackhumbert** [comment]
Yeah, the spacing is located here: https://github.com/jackhumbert/in_world_navigation/blob/main/src/redscript/in_world_navigation/InWorldNavigation.reds#L29 - everything else is in the effect, particle, or material, which are all kind of hard to configure at the moment. My plan is to just have a couple different options, instead of allowing full configuration of each element.

**Comment by Thortok** [comment]
In the current version, the fade distance when on foot is too far, imo. The previous version was helpful for navigating staircases and such, in this version it's basically gone on staircases. It's a good fade distance for being in a vehicle though.

One thing I wanted to try was like a distance where you set it as completely invisible and a distance where you set it as completely opaque. On I'd think like a distance of somewhere from 1-10 for the invisible, and 100 for the opaque.

What does 'distanceToAnimate' do?

**Comment by bknie1** [comment]
Heya-

I spent the evening trying to integrate [Native Settings UI](https://github.com/justarandomguyintheinternet/CP77_nativeSettings) into your mod. The result was the below script to begin to integrate a mods menu option, an In Game Navigation mods menu tab, and three categories (Spacing, Length, and Distance) to modify the existing values in your setup. However,  I'm stumped when it comes to including the Native Settings UI mod as a dependency. I think the below will work if someone can help me bridge the two mods.

I've omitted unchanged lines in this snippet. I've also included verbose comments that should be deleted later. I'm not familiar with the Red language and the Native UI mod looks like Lua. I thought, perhaps, I would be able to modify the Reds file where the measurement data points were. I'm unsure if this was the right approach. This is my first time messing with CP2077 mod code.

```
let nativeSettings = GetMod("nativeSettings"); // This is the dependency. But I don't think this is the correct way to load it

  public func Setup(player: ref<GameObject>) -> Void {
    this.player = player; // Pre-existing but followed by ...
    SetRenderDefaultValues(); // Spacing, length, distances. Called once on setup, bound to 'Defaults' mod button
    
    // ...

    let addedModTab = AddNativeUI(); // Mod menu setup abstracted to keep it tidy
  }

  private func SetRenderDefaultValues() -> void {
    // Called twice: Setup, on 'Reset to defaults' mod tab button
    this.spacing = 5.0; // meters
    this.maxPoints = 200;
    this.distanceToAnimate = 50.0;
    this.distanceToFade = 25.0;
  }

  private func AddNativeUI() -> bool {
    // Setup in-game Native Settings UI 'Mods' menu tab
    do {
      // == MENU TAB =========================================================
      nativeSettings.addTab("/holonav", "In World Navigation");

      // == CATEGORIES ========================================================
      // Spacing: Between indicators
      nativeSettings.addSubcategory("/holonav/spacing", "Spacing");
      nativeSettings.addRangeFloat(
        "/holonav/spacing", "Space Between Markers", "Markers are spaced a specific distance apart (m)", 
        5, 50, 1.0, "%.2f", 50, 1,
        function(value)
 	          this.spacing = value;
      end);
      
      // Length: Maximum number of indicators
      nativeSettings.addSubcategory("/holonav/length", "Length");
      nativeSettings.addRangeInt(
        "/holonav/length", "Path", "The marker path is rendered for a specific distance (m)",
         1, 200, 1, 50, 25, 
        function(value)
          this.maxPoints = value;
      end)

      // Distance: Animation and Fade
      nativeSettings.addSubcategory("/holonav/distance", "Distance");
      nativeSettings.addRangeFloat(
        "/holonav/distance", "Marker Animation", "Markers are animated for a specific distance (m)",
         25, 200, 1.0, "%.2f", 50, 1, 
         function(value)
 	          this.distanceToAnimate = value;
      end);
      nativeSettings.addRangeFloat(
        "/holonav/distance", "Marker Fade", "Markers begin to fade after a specific distance (m)",
         25, 200, 1.0, "%.2f", 50, 1, 
         function(value)
            this.distanceToFade = value;
      end);

      // Reset to default config
      nativeSettings.addButton(
        "/holonav/", "ResetButton", "Resets values to their default configuration", "Default", 45, 
        function() {
          SetDefaultValues();
        }
      end);
    }
    catch {
        // Failed to setup in-game menu tab
        print("ERROR - Failed to load Native UI 'Mods' menu tab for In World Navigation. Continuing ...");
        return false;
    }

      return true;
  }
```

**Comment by jackhumbert** [comment]
Nice! I'm actually working on porting the Native Settings functionality to redscript/red4ext, so that these sort of settings can work with more parts of the game (and avoid additional dependencies), but this is a great start to the configuration part of that.

**Comment by bknie1** [comment]
> Nice! I'm actually working on porting the Native Settings functionality to redscript/red4ext, so that these sort of settings can work with more parts of the game (and avoid additional dependencies), but this is a great start to the configuration part of that.

Pardon my enthusiasm :) I really like where your mod is going. As a player that tends to turn off map systems, for the sake of immersion, I think your mod can provide the happy medium of well spaced holo-signage and immersion. Also, just to be transparent, if you're looking for a little less collaboration, or prefer this to be a solo project, I will completely understand.

With that said, here are my follow up thoughts after sleeping on it:

## What's "Done"

The variables are there and the 'Mods' UI config is good to go.

## To Do

- How do we bridge the gap between your mod and this would-be dependency?
- There is the potential for a lightweight version of the mod and a UI version of the mod.

### In-World Navigation: Lite Edition

You mentioned wanting to minimize dependencies and I think there's value in having a 'lite' Release with a Readme that points users to those const values. Essentially, what you already have could be treated as its own MVP for users that don't want the bloat or more dependencies. Or, better yet, continue packing in those dependencies as an all-in-one for users on Nexus etc.

### In-World Navigation: UI Edition

This would require a custom interface or by leveraging the existing Native Settings UI module:

## Design Questions

- Do we abstract away the load logic to a more global interface or try to incorporate it into your C++ project?
 
### Global Interface and Dependency Solution
- Where would Native UI gets plugged into RedText/CET/some interface? Which is best? How do other mods do it? (e.g. Metro)

### Mod Level Custom Solution
- How do we import a module written in another language? We can't simply import a Header because it was written in... Lua, I think. The core of your mod is C++ and I see you using the Red language .reds to interact with it. How would we plug a .lua project into that?

- Once it's compiled, at run time, can we use the settings UI to modify the values in memory?

- Short term solution: Users that change settings can, at least, write these values to runtime memory; even if the changes aren't saved/persistent

- Long term solution: Once values are changed in memory, those values need to be saved to a state file; likely as plain text values that can be modified outside the game. At this point, this infrastructure likely doesn't exist because the values aren't changing. Fortunately, I already know how to write this according to the UI dependency: We would instantiate a nativeSettings button that triggers an anonymous function that writes these values, by property name and value, to a file. On load, if state file exists, we would gracefully try to detect and read an existing state file. If the load fails, the mod can load default values.

## Thanks for coming to my Brand Talk

If you can point me in the right direction I'd be interested in testing what I come up with against what you come up with. I'm just not sure how to import/plug in this mod so that it can write settings to yours. I'm unsure if I was even working in the right location. I haven't really contributed to a random open source project on GitHub yet so I'd love to see it through. Again, if this is something you would prefer to solo I completely understand.

I also totally have a day job to do but this is super fun and I really appreciate you making the mod. Even as it is, without any additional work, it is immensely useful. Thank you @jackhumbert !

**Comment by bknie1** [comment]
P.S. here's my "minimalist" config for your mod in the mean time:

```
    this.player = player;
    this.spacing = 200.0; // Spaced farther apart / off in the distance for a quick hint but not in your face
    this.maxPoints = 5; // You may really only need 1-2, honestly
    this.distanceToAnimate = 0.0; // Turns of animation; was a bit distracting for me
    this.distanceToFade = 200.0; // Tinkering with this still. Trying to get the arrows to fade sooner/while I'm still farther away
```

**Comment by jackhumbert** [comment]
Implemented in [v0.0.3](https://github.com/jackhumbert/in_world_navigation/releases/tag/v0.0.3) with [Mod Settings](https://github.com/jackhumbert/mod_settings).
